### PR TITLE
export low-level baseXX encoder and decoder from dnsext-types

### PR DESCRIPTION
**Pull request by khibino**
Want to fix #49 by applying `decodeBase32Hex` instead of base32 package

### CODE REVIEW

File: dnsext-types/DNS/Types/Base32Hex.hs [hunk 1]
```diff
@@ -51,3 +55,82 @@ encode bs =
...
+decode :: ByteString -> Either String ByteString
+decode bs = do
+  let len = (5 * BS.length bs) `div` 8
+  cs <- fmap chunks8 . mapM fromHex32 $ BS.unpack bs
+  return $ BS.pack $ A.elems $ A.runSTUArray $ do
+    a <- A.newArray (0 :: Int, len-1) 0
+    go cs a 0
+  where
+    fromHex32 w
+      {- '0' <= w <= '9' -}
```
Comment (kazu-yamamoto) [inline dnsext-types/DNS/Types/Base32Hex.hs:68]: Remove this comment and use Data.Word8._0, etc.

File: dnsext-types/DNS/Types/Base32Hex.hs [hunk 2]
```diff
@@ -51,3 +55,82 @@ encode bs =
...
+      where
+        chunk s
+          | null hd    =  Nothing
+          | otherwise  =  Just (hd, tl)
+          where (hd, tl) = splitAt 8 s
+
+    store8 a i v = A.writeArray a i v
+
+    stores ss a n f0 f1 f2 f3 f4 f5 f6 f7 = do
+      let w0 = f0 `shiftL` 3 .|. f1 `shiftR` 2
```
Comment (kazu-yamamoto) [inline dnsext-types/DNS/Types/Base32Hex.hs:86]: Replace shiftL with unsafeShiftL if this operation is safe.

**Review by kazu-yamamoto — COMMENTED** [review]
This RP could be improved.

**Comment by khibino** [comment]
Perhaps an interface via `Opaque` would be considered better?

https://github.com/kazu-yamamoto/dnsext/compare/base-encoding..base-encoding-opaque

**Comment by kazu-yamamoto** [comment]
Yes. Would you make additional changes?

**Comment by khibino** [comment]
done. additional changes are pushed.

**Review by kazu-yamamoto — APPROVED** [review]
Now LGTM

**Comment by kazu-yamamoto** [comment]
Merged.

**Pull request merged**